### ISSUE TITLE
Keeping DDP override in sync with upstream torch

### ISSUE THREAD

**Issue by edenlightning** [PyTorchLightning/pytorch-lightning#4630]
From @ananthsub:
how should Lightning keep its [DDP override](https://github.com/PyTorchLightning/pytorch-lightning/blob/3d202f9ecc4137b08cb5b1ac15af276456fcfaaf/pytorch_lightning/overrides/data_parallel.py#L154-L210) in sync with the upstream torch [DistributedDataParallel](https://github.com/pytorch/pytorch/blob/d478605dec65a746d41506b23693d6013bfa11b2/torch/nn/parallel/distributed.py#L646-L697)? these implementations have now diverged. I think this leads to performance degradations with Lightning + gradient accumulations, since the require_backward_grad_sync attribute isn't checked before the backwards pass

**Comment by edenlightning** [comment]
@awaelchli @tchaton

**Comment by ananthsub** [comment]
@pritamdamania87 suggested this workaround. Instead of overriding DDP, we can wrap the LightningModule in another nn.Module. This wrapper module will define it's `forward` function to call the `LightningModule`'s `*_step` function, depending on the `training/testing` flags. 

Then when we wrap this module in DDP, we can rely on the wrapper's `forward` to do the right thing, and we don't have to worry about keeping the overrides in sync

**Comment by pritamdamania87** [comment]
As @ananthsub I'd suggest always calling DDP.forward and not relying on the internals of DDP. The current implementation in LightningDistributedDataParallel could break if we make changes in DDP and the reducer. As an example, mmcv was doing something similar and broke when we refactored some code related to DDP and reducer: https://github.com/open-mmlab/mmcv/issues/636.

**Comment by awaelchli** [comment]
@ananthsub I think this could work, nice idea! And I guess we should unwrap the model when passing it to the callbacks etc.

**Comment by awaelchli** [comment]
solved by the linked PR
I will open a new issue for a similar refactor to DP.
